executor/src/witgen/data_structures/mutable_state.rs

```diff
@@ -0,0 +1,86 @@
+use std::{
+    cell::RefCell,
+    collections::{BTreeMap, HashMap},
+};
+
+use powdr_number::FieldElement;
+
+use crate::witgen::{
+    machines::{KnownMachine, Machine},
+    rows::RowPair,
+    EvalError, EvalResult, QueryCallback,
+};
+
+/// The container and access method for machines and the query callback.
+/// The machines contain the actual data tables.
+/// This struct uses interior mutability for accessing the machines.
+pub struct MutableState<'a, 'b, T: FieldElement, Q: QueryCallback<T>> {
+    machines: Vec<RefCell<KnownMachine<'a, T>>>,
+    identity_to_machine_index: BTreeMap<u64, usize>,
```

Comment:
```suggestion
    identity_to_machine: BTreeMap<u64, RefCell<KnownMachine<'a, T>>>,
```
Why not this?

Reply:
take_witness_col_values uses the given order of machines. Also If we do it like that, then multiple identities calling the same machine will lead to a copy of the machine, which we don't want to have.

Reply:
Two good reasons!